src/core/tsi/s2a/frame_protector/s2a_frame_protector.cc

```diff
@@ -0,0 +1,246 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <openssl/ssl3.h>
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/tsi/alts/handshaker/alts_tsi_utils.h"
+#include "src/core/tsi/s2a/frame_protector/s2a_frame_protector.h"
+#include "src/core/tsi/s2a/record_protocol/s2a_crypter.h"
+#include "src/core/tsi/s2a/record_protocol/s2a_crypter_util.h"
+#include "src/core/tsi/s2a/s2a_constants.h"
+
+typedef struct s2a_zero_copy_grpc_protector {
+  tsi_zero_copy_grpc_protector base;
+  s2a_crypter* crypter;
+  size_t max_protected_frame_size;
+  size_t max_unprotected_data_size;
+  grpc_slice_buffer protected_sb;
+  grpc_slice_buffer protected_staging_sb;
+  grpc_slice_buffer unprotected_staging_sb;
+  size_t current_frame_size;
+} s2a_zero_copy_grpc_protector;
+
+/** --- tsi_zero_copy_grpc_protector methods implementation. --- **/
+
+static tsi_result s2a_zero_copy_grpc_protector_protect(
+    tsi_zero_copy_grpc_protector* self, grpc_slice_buffer* unprotected_slices,
+    grpc_slice_buffer* protected_slices) {
+  if (self == nullptr || unprotected_slices == nullptr ||
+      protected_slices == nullptr) {
+    gpr_log(GPR_ERROR, kS2AProtectNullptr);
+    return TSI_INVALID_ARGUMENT;
+  }
+  s2a_zero_copy_grpc_protector* protector =
+      reinterpret_cast<s2a_zero_copy_grpc_protector*>(self);
+  char* error_details = nullptr;
+
+  while (unprotected_slices->length > protector->max_unprotected_data_size) {
+    grpc_slice_buffer_move_first(unprotected_slices,
+                                 protector->max_unprotected_data_size,
+                                 &(protector->unprotected_staging_sb));
+    grpc_status_code status = s2a_protect_record(
+        protector->crypter, SSL3_RT_APPLICATION_DATA,
+        &(protector->unprotected_staging_sb), protected_slices, &error_details);
+    if (status != GRPC_STATUS_OK) {
+      gpr_log(GPR_ERROR, "Failed to protect record: %s", error_details);
+      gpr_free(error_details);
+      return alts_tsi_utils_convert_to_tsi_result(status);
```

Comment:
are you going to refactor this util function into a shared tsi_util library?

Reply:
Yes, that's correct. See my response to your 2 high-level comments.